Task4

quperino opened the pull request.

Commit:
Решение задачи 4

quperino committed:
Исправил

PuccaGTS commented on src/main/java/com/walking/intensive/chapter1/task4/Task4.java:
```diff
@@ -24,18 +24,30 @@
+            double x1 = (-b - Math.sqrt(discriminant)) / (2 * a);
+            double x2 = (-b + Math.sqrt(discriminant)) / (2 * a);
+            if (x1 > x2) {
+                s = "Количество решений: 2. Корни: " + x2 + "; " + x1;
```
Пробел между точкой с запятой не нужен, в условии его нет

quperino replied:
Исправил

PuccaGTS commented on src/main/java/com/walking/intensive/chapter1/task4/Task4.java:
```diff
@@ -24,18 +24,30 @@
 
     static String solveEquation(double a, double b, double c) {
-        //        Место для вашего кода
+        double discriminant = (Math.pow(b, 2)) - (4 * a * c);
```
Кажется, что в некоторых случаях нет необходимости считать дискриминант

quperino replied:
Исправил

PuccaGTS commented on src/main/java/com/walking/intensive/chapter1/task4/Task4.java:
```diff
@@ -24,18 +24,30 @@
 
     static String solveEquation(double a, double b, double c) {
-        //        Место для вашего кода
+        double discriminant = (Math.pow(b, 2)) - (4 * a * c);
```
Если передать значения 0, 0, -10, то получится уравнение вида:
0x^2 + 0x -10 = 0
-10 = 0 - количество решение тут 0, а у Вас Дискриминант есть и корни есть:)

quperino replied:
Исправил

PuccaGTS commented on src/main/java/com/walking/intensive/chapter1/task4/Task4.java:
```diff
@@ -24,18 +24,30 @@
+        if (discriminant < 0) {
+            s = "Количество решений: 0.";
+        } else if (discriminant == 0) {
+            s = "Количество решений: 1. Корень: " + (-b / 2 * a);
```
При значениях 0, 0, 0 - хотелось бы получить "Бесконечное множество решений"
А у Вас - один корень)

quperino replied:
Исправил

PuccaGTS commented on src/main/java/com/walking/intensive/chapter1/task4/Task4.java:
```diff
@@ -24,18 +24,30 @@
+            s = "Количество решений: 0.";
+        } else if (discriminant == 0) {
+            s = "Количество решений: 1. Корень: " + (-b / 2 * a);
+        } else if (discriminant > 0) {
```
А вот сюда попадает кейс, когда a = 0, например:
0, -2, 2 или 0, 20, -40
У них один корень, т.к. уравнение получается линейное, то по Вашему кода Дискриминант положительный и корни вычисляются неверно.

quperino replied:
Исправил

PuccaGTS reviewed (changes requested):
Давайте поправим

quperino committed:
Исправил, согласно комментариям. Теперь условия поставлены без опоры на дискриминант, а высчитывается он только по необходимости, т.е. в случае, который не встретился в первых трех проверках

PuccaGTS commented on src/main/java/com/walking/intensive/chapter1/task4/Task4.java:
```diff
@@ -24,18 +24,32 @@
+        } else if (a == 0 && c != 0) {
+            s = "Количество решений: 1. Корень: " + (-c / b);
+        } else {
+            double discriminant = (Math.pow(b, 2)) - (4 * a * c);
```
Действительные корни отсутствуют при отрицательном дискриминанте

quperino replied:
Исправил

PuccaGTS commented on src/main/java/com/walking/intensive/chapter1/task4/Task4.java:
```diff
@@ -24,18 +24,32 @@
+        } else if (a == 0 && c != 0) {
+            s = "Количество решений: 1. Корень: " + (-c / b);
+        } else {
+            double discriminant = (Math.pow(b, 2)) - (4 * a * c);
```
Если дискриминант равен 0, то корень всего один. а у вас получается два -0.0 и 0.0

quperino replied:
Исправил

PuccaGTS reviewed (changes requested):
Давайте поправим

quperino committed:
Исправил условия, согласно комментариям

PuccaGTS commented on src/main/java/com/walking/intensive/chapter1/task4/Task4.java:
```diff
@@ -24,18 +24,32 @@
+            s = "Бесконечное множество решений.";
+        } else if (b * b < 4 * a * c) {
+            s = "Количество решений: 0.";
+        } else if (b * b == 4 * a * c) {
```
Теперь под это условие попадает случай, который мы с Вами рассматривали в самом начале, когда передаем 0, 0, -10. Ждали Количество решений: 0., а получили Количество решений: 1. Корень: Infinity

PuccaGTS replied:
А сюда не попал кейс, когда это линейное уравнение и нужно считать так, как Вы написали, но из-за того, что условие неправильно написано, случай не обрабатывается, пример:
```
Правильное решение:
0x^2 - 2x +8 = 0
-2x + 8 = 0
x = 4.0

Ваше решение:
if (4 == 4 * 0 * 8 = 0) false
и идем дальше в блок else, чтобы считать два корня
```
В итоге ожидаем:
Количество решений: 1. Корень: 4.0
Получаем:
Количество решений: 2. Корни: NaN;Infinity

PuccaGTS commented on src/main/java/com/walking/intensive/chapter1/task4/Task4.java:
```diff
@@ -24,18 +24,32 @@
+        } else if (b * b < 4 * a * c) {
+            s = "Количество решений: 0.";
+        } else if (b * b == 4 * a * c) {
+            s = "Количество решений: 1. Корень: " + (-c / b);
```
Не всегда всегда дискриминант равный нулю значит, что это не квадратное уравнение и нужно корни считать только использую линейную функцию, например:
```
x^2 - 2x + 1 = 0
x^2 - 10x + 25 = 0
```
и т.д.

PuccaGTS replied:
Плюс сюда же кейс x^2 = 0;
Ожидаем: Количество решений: 1. Корень: 0.0
Получаем: Количество решений: 1. Корень: NaN - т.к. делим 0 на 0

PuccaGTS reviewed (changes requested):
Давайте поправим

quperino committed:
Полностью переписал код, прошлые комментарии к старому коду учел

PuccaGTS commented on src/main/java/com/walking/intensive/chapter1/task4/Task4.java:
```diff
@@ -24,18 +24,37 @@
+            s = "Количество решений: 1. Корень: " + -c / b;
+        } else if (b * b == 4 * a * c) {
+            s = "Количество решений: 1. Корень: " + -b / 2 * a;
+        } else if (a > 0 && c < 0 || a < 0 && c > 0) {
```
Под это условие попало уравнение:
`x^2 + 4x - 45 = 0` - валидное уравнение, которое имеет корни -9.0 и 5.0
Ваша программа выдает NaN;NaN, т.к. вычисляет корень из отрицательного числа

PuccaGTS replied:
Есть и другие примеры, похожие, которые также попадают в это условие

quperino replied:
Добавил уточняющее условие && b == 0 два раза. Также убрал из верхних конструкций проверку на значение дискриминанта, вынес его в последнее условие. Switch case не воспользовался, так как пока что не совсем понял, как в нем можно будет реализовать конструкцию, подобной той, которую я написал через if-else

PuccaGTS reviewed (changes requested):
Давайте попробуем оптимизировать Ваше решение, сейчас условные блоки выглядят очень громоздко и сложно понять, какое условие работает, предлагаю пойти таким путем:
1. Сперва посмотреть на коэффициенты и сделать вывод, если это возможно.
2. Если ответа в первом пункте не было, то посчитать дискриминант и сделать вывод на основе дискриминанта, важно - дискриминант лучше всего посчитать один раз, а потом переиспользовать, чтобы избежать повторения кода, как у Вас:
`b * b == 4 * a * c`
`b * b - 4 * a * c`
Вы, по-сути, вычисляете одно и то же, но можно это сделать один раз и потом переиспользовать.

Ещё можно воспользоваться структурой switch case для множественного выбора, чтобы избежать большого количества if-else if- else if, так код будет чище.

Вы молодец, осталось совсем немного:)

quperino committed:
Добавил уточняющее условие && b == 0 в конструкцию, теперь одно из условий выглядит, как:
```
else if (a > 0 && b == 0 && c < 0 || a < 0 && b == 0 && c > 0) {
    double x1 = Math.sqrt(c / a);
    double x2 = -Math.sqrt(c / a);
    s = "Количество решений: 2. Корни: " + Math.min(x1, x2) + ";" + Math.max(x1, x2);
}
```

Проверку на значение дискриминанта вынес в последнее условие в конструкции
